core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java

```diff
@@ -260,8 +259,16 @@ public long readVLong() throws IOException {
             return i;
         }
         b = readByte();
-        assert (b & 0x80) == 0;
-        return i | ((b & 0x7FL) << 56);
+        i |= ((b & 0x7FL) << 56);
+        if ((b & 0x80) == 0) {
+            return i;
+        }
+        b = readByte();
+        if (b != 0 && b != 1) {
+            throw new IOException("Invalid vlong (" + Integer.toHexString(b) + " << 63) | " + Long.toHexString(i));
+        }
+        i |= (b & 0x7FL) << 63;
```

Comment:
Oh, I'm fine with symmetry, I just wanted to make sure that I was reading it correctly.